### DISCUSSION TITLE
Cannot connect srsUE to gNB through ZMQ. Gets stuck on RRC Connected.

### DISCUSSION THREAD

**Discussion by fabio-patrizi** [srsran/srsRAN_Project#92] [General Help] [answered]
Hi all,
I'm having troubles in connecting the srsUE to the gNB via ZMQ. 

I've followed the steps in the  [tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#srsran-gnb-with-srsue). I have built open5gs, srsRAN_Project and srsRAN_4G from sources on Ubuntu 22.04.2 LTS (actually, I'm on a Mac M1 Max).  Each of them seems to work fine and apparently the gNB and open5gs connect with each other. However, when the srsUE seems to fail in connecting. Details follow.

gNB  commit is ce86e4770
srsUE commit is 921f17484

(Expected behavior)
As per the tutorial, the srsUE output should end by something as follows:
...
RRC Connected
PDU Session Establishment successful. IP: 10.45.10.2
RRC NR reconfiguration successful.

(Observed behavior)
However, I get the following output:

Active RF plugins: libsrsran_rf_uhd.so libsrsran_rf_zmq.so
Inactive RF plugins: 
Reading configuration file ./srsRAN_4G/build/ue_zmq.conf...

Built in Release mode using commit 921f17484 on branch master.

Opening 1 channels in RF device=zmq with args=tx_port=tcp://127.0.0.1:2001,rx_port=tcp://127.0.0.1:2000,base_srate=11.52e6
Supported RF device list: UHD zmq file
CHx base_srate=11.52e6
Current sample rate is 1.92 MHz with a base rate of 11.52 MHz (x6 decimation)
CH0 rx_port=tcp://127.0.0.1:2000
CH0 tx_port=tcp://127.0.0.1:2001
Current sample rate is 11.52 MHz with a base rate of 11.52 MHz (x1 decimation)
Current sample rate is 11.52 MHz with a base rate of 11.52 MHz (x1 decimation)
Waiting PHY to initialize ... done!
Attaching UE...
Random Access Transmission: prach_occasion=0, preamble_index=0, ra-rnti=0x39, tti=334
Random Access Complete.     c-rnti=0x4601, ta=0
RRC Connected

In particular, I do not get the lines: 

PDU Session Establishment successful. IP: 10.45.10.2
RRC NR reconfiguration successful.

which indicate successful connection (according to the tutorial). 

When I test the connection with ping, it doesn't work.

gNB output is: 
Lower PHY in executor blocking mode.

--== srsRAN gNB (commit ce86e4770) ==--

Connecting to AMF on 127.0.0.5:38412
Available radio types: uhd and zmq.
Cell pci=1, bw=10 MHz, dl_arfcn=368500 (n3), dl_freq=1842.5 MHz, dl_ssb_arfcn=368410, ul_freq=1747.5 MHz

==== gNodeB started ===
Type <t> to view trace

Configuration files are attached:
[gnb_zmq.txt](https://github.com/srsran/srsRAN_Project/files/11680309/gnb_zmq.txt)
[ue_zmq.conf.txt](https://github.com/srsran/srsRAN_Project/files/11680315/ue_zmq.conf.txt)

EDIT: forgot to attach open5gs config files
[amf.yaml.txt](https://github.com/srsran/srsRAN_Project/files/11682492/amf.yaml.txt)
[upf.yaml.txt](https://github.com/srsran/srsRAN_Project/files/11682493/upf.yaml.txt)

As an aside, the field rf_driver in gnb_zmq.txt does not seem to be correctly parsed. Indeed I get the following error when running the gNB:
INI was not able to parse rf_driver.++
Run with --help for more information.

EDIT: 
Also using "ss_type: common" in gnb_zmq.txt does not work and produces the following error:
INI was not able to parse cell_cfg.ss_type
Run with --help for more information.

For this reason, after inspecting the config files provided with the release, I switched to ru_sdr. If I do so, it works fine. This is commented inline in the gNB configuration file.

To reproduce the issue, just run the open5gs, followed by the gNB, then by the srsUE. If needed, I can provide the specific command lines used, as well as the logs. 

Thank you in advance for any help you can provide.
Best!

**Comment by oguztudo** [comment]
I encountered the same issue when i forgot to add ue1
Check with :
**sudo ip netns list** 

if the list is empty try:
**sudo ip netns add ue1**

**Reply by fabio-patrizi** [reply]
Hi, 
Thanks a lot for the reply. Yes, I added ue1. In fact, I'm using a script which does it every time before srsUE is launched (and after the first time itś always there).

One thing Iḿ not sure about, which I forgot to post in the original discussion, is the subscriber configuration. In particular, I do not know how to set AMF, and I left it assigned to 8000. I attach the screenshots below.  See that in the picture, AMF is 9000 because I was trying a different configuration, but even leaving it to the default (8000) it doesn't work.

BTW, did you have the same issue with rf_driver vs ru_sdr? 
Could you share your gNB and ue configuration files? It might be helpful to compare with mine.

Thanks a lot!

![Screenshot from 2023-06-07 17-56-23](https://github.com/srsran/srsRAN_Project/assets/55386576/cfcb228e-14fd-400e-b40e-6a9c7bf7efba)
![Screenshot from 2023-06-07 17-57-42](https://github.com/srsran/srsRAN_Project/assets/55386576/d66cde84-4e19-4f67-ae0c-65fe93d91ba9)
![Screenshot from 2023-06-07 17-59-04](https://github.com/srsran/srsRAN_Project/assets/55386576/b6ec6f90-5e37-486c-886b-545b7ddf3fff)

**Reply by oguztudo** [reply]
Yes I had the same problem and looking again over everything i realised you commented out Search Space type common. As far as i know this should be set to common, but i also know that it won't work with this setting right now, due to an error. So you most likely have to wait for a fix by the team of srsRAN if you want to use the latest version.
You could also pull the 23.5 version of srsRAN project from git and make a build with that:
https://github.com/srsran/srsRAN_Project/releases/tag/release_23_5
It's not the latest code but you will be able to use ZMQ here. 
Hope that helps :)

**Reply by fabio-patrizi** [reply]
Thanks for the help. I'll try with v. 23.5. For now it should be OK. I'll be waiting for the fix.
Cheers.

**Reply by ismagom** [reply]
Hi @fabio-patrizi and @bobson51 , the option `--ss_type` is now part of the `pdcch` section, eg: 

```
cell_cfg: 
  pdcch: 
    ss_type: common
```

If you see this error again, use the command line help, eg: 

```
$ apps/gnb/gnb --help
srsGNB application
Usage: apps/gnb/gnb [OPTIONS] [SUBCOMMAND]

Options:
  -h,--help                   Print this help message and exit
  -v,--version                Display program version information and exit
  -c                          Read config from file
  --gnb_id UINT [411]         gNodeB identifier
  --gnb_id_bit_length UINT:INT in [22 - 32] [ ] 
                              gNodeB identifier length in bits
  --ran_node_name TEXT [srsgnb01] 
                              RAN node name
  --cells TEXT ...            Sets the cell configuration on a per cell basis, overwriting the default configuration defined by cell_cfg
  --qos TEXT ...              qos

Subcommands:
  log                         Logging configuration
  pcap                        PCAP configuration
  amf                         AMF parameters
  cu_cp                       CU-CP parameters
  ru_sdr                      SDR Radio Unit configuration
  ru_ofh                      Open Fronthaul Radio Unit configuration
  cell_cfg                    Default cell configuration
  expert_phy                  Expert physical layer configuration
  test_mode                   Test mode configuration
```
then if you type: 

```
$ apps/gnb/gnb cell_cfg pdcch --help
PDCCH parameters
Usage: apps/gnb/gnb cell_cfg pdcch [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  --ss_type TEXT:{common,ue_dedicated} [ue_dedicated] 
                              SearchSpace type for UE data
  --dci_format_0_1_and_1_1 BOOLEAN [1] 
                              DCI format to use for UE data
```

**Reply by oguztudo** [reply]
Thanks for the clarification @ismagom and sorry for the misinformation @fabio-patrizi .

**Reply by fabio-patrizi** [reply]
Hi @ismagom,
thanks for the additional info. Actually, I have moved this discussion to the issue section. Sorry, I forgot to mention this. I'll close this discussion to avoid confusion. 
Apparently, the tutorial has now been updated with the correct configuration files. I do not get the errors related to "rf_driver" and "ss_type: common" but I still can't manage to connect the UE to the gNB. I must have introduced some error in the config files.
Thanks!

@bobson51: no problem!

**Reply by SadeghKrmi** [reply]
@fabio-patrizi, I have the same issue, I got stuck at the RRC Connected, srsUE is not able to get IP address, have you found any workaround?

**Reply by herlesupreeth** [reply]
Please use the latest commits in srsUE repo. The issue has been fixed.

**Reply by SadeghKrmi** [reply]
@herlesupreeth thanks for the response, shall I use srsUE from srsRAN_4G or independent srsUE?

**Reply by herlesupreeth** [reply]
srsUE from srsRAN_4G

**Reply by SadeghKrmi** [reply]
@herlesupreeth I run the git pull and make it again for the srsRAN-4G, but still, I stuck at RRC Connected, I am not receiving any IP from UPF.

**Reply by ismagom** [reply]
Can you attach logs setting `all_level = info` please? without logs we can't see anything

**Reply by SadeghKrmi** [reply]
@ismagom Sure, thanks for supporting, log for ue and gNB are as follows 
[ue.log](https://github.com/srsran/srsRAN_Project/files/11917059/ue.log)
[gnb.log](https://github.com/srsran/srsRAN_Project/files/11917080/gnb.log)

**Reply by herlesupreeth** [reply]
I see from your ue.log the following

> 2023-06-30T14:06:58.088603 [UE     ] [I] Built in Release mode using commit 49f1edd6f on branch master.

But the latest commit in master branch of srsRAN_4G is fa56836b14dc6ad7ce0c3484a1944ebe2cdbe63b. Notice that they do not match.

**Reply by SadeghKrmi** [reply]
@herlesupreeth Hey, I used latest commit, but I am facing same issue, It got stuck at RRC Connected, without receiving IP. I am attaching new logs, I'd appreciate your opinion.

[gnb.log](https://github.com/srsran/srsRAN_Project/files/11967688/gnb.log)

[ue.log](https://github.com/srsran/srsRAN_Project/files/11967686/ue.log)

**Reply by herlesupreeth** [reply]
I took a look at the logs but couldn't find anything that could give a hint about the issue. Can you please increase the log level to debug for following layers under `log` configuration for gNB and also enable ngap pcap and send both the logs and ngap pcap. Thanks
```

  phy_level: debug
  mac_level: debug
  rrc_level: debug
```

**Reply by SadeghKrmi** [reply]
Sure, please find the logs in below.

[ue.log](https://github.com/srsran/srsRAN_Project/files/11979804/ue.log)
[gnb.log](https://github.com/srsran/srsRAN_Project/files/11979805/gnb.log)
[gnb_ngap.zip](https://github.com/srsran/srsRAN_Project/files/11979809/gnb_ngap.zip)

**Comment by herlesupreeth** [comment, marked as answer]
@SadeghKrmi I see that Core Network is rejecting your UE registration request so its a issue in your Core Network configuration. I would suggest to check whether you have correctly entered UE's SIM details in the Core Network and also check the MNC and MCC.

![image](https://github.com/srsran/srsRAN_Project/assets/4115839/f8b8b8e3-502f-4ab2-8167-9344a52e62e0)

**Reply by SadeghKrmi** [reply]
@herlesupreeth thanks for the response, I installed Mongodb on ubuntu 22.04 earlier, but seems because I am using virtualization and CPU is not supporting features needed by Mongodb, it fails to startup the db,  I switched to ubuntu 20.04 to be able to install Mongodb of version 4, then this issue is fixed, now I am receiving "PLMN not allowed" error which I am trying to solve.

![image](https://github.com/srsran/srsRAN_Project/assets/5988663/72408fbd-f4dd-4323-a860-450ac212f480)

**Reply by herlesupreeth** [reply]
Its the same error as before. I would suggest checking your UE details entered in mongo db / WebUI of open5gs

**Reply by SadeghKrmi** [reply]
@herlesupreeth thanks, now it is fixed, I inserted the wrong UE imsi.

**Comment by KianooshSepas** [comment]
Hi all,

I've followed the [srsRAN installation guide](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/installation.html) and successfully set up Open5GS as the core network. However, when I try to run the gNB using the following command:

bash

./gnb -c ./gnb_zmq.yaml

I encounter the following error:

sql

--== srsRAN gNB (commit ) ==--

INI was not able to parse amf.++
Run with --help for more information.

Could someone help me resolve this issue? It seems to be related to the amf configuration in the gnb_zmq.yaml file, but I haven't been able to fix it.
and this I just download the files from here https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#zeromq-based-setup

Thanks in advance!

**Reply by pgawlowicz** [reply]
Thanks for pointing this issue out. Please modify the `gnb_zmq.yaml` file as follows:

remove:
```
amf:
  addr: 10.53.1.2                  # The address or hostname of the AMF.
  bind_addr: 10.53.1.1             # A local IP that the gNB binds to for traffic from the AMF.
```

add 
```
cu_cp:
  amf:
    addr: 10.53.1.2
    port: 38412
    bind_addr: 10.53.1.1
    supported_tracking_areas:
      - tac: 7
        plmn_list:
          - plmn: "00101"
            tai_slice_support_list:
              - sst: 1
```
